core/discovery/interface.go

```diff
@@ -29,8 +29,22 @@ type ProposalRegistry interface {
 	UnregisterProposal(proposal market.ServiceProposal, signer identity.Signer) error
 }
 
-// ProposalFinder allows to search proposals by specified params
+// ProposalReducer proposal match function
+type ProposalReducer func(proposal market.ServiceProposal) bool
+
+// ProposalMatcher defines interface to fetch currently active service proposals from discovery by match function
+type ProposalMatcher interface {
+	GetProposal(id market.ProposalID) (*market.ServiceProposal, error)
+	MatchProposals(match ProposalReducer) ([]market.ServiceProposal, error)
+}
+
+// ProposalFilter defines interface with proposal match function
+type ProposalFilter interface {
+	Matches(proposal market.ServiceProposal) bool
```

Comment:
can we just define this as a func type instead of an interface?

Reply:
`ProposalReducer` is function, and `ProposalFilter` interface to record which filter was used in tests

Reply:
Anyways, I improved it now by moving interfaces to caller